Avatar item token improvement suggestion

shaokun11 opened the pull request:
Main change is update avatar item structure  and updated functions to match the changes

shaokun11 committed:
Main change is update avatar item structure  and updated functions to match the changes

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
 
   AvatarItemService internal itemService;
-
+  
```
lint: remove extra space

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
+  function getTokenInfo(uint256 _tokenId)
+    external 
+    view 
+    returns(address, address, bool, int16, uint256[4] _attr1, uint8[5] _attr2) {
```
You've named the arrays "_attr1", "_attr2", can you also name the return values of other fields to make it more readable?

shaokun11 replied:
I can't achieve it for there are so many variable,it will cause stake to deep error

calvinmd replied:
Oh, yeah, that weird stack too deep error! Gotcha!

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
+    int16[] _probabilities,  // one time
+    address[] _addresses,    // two times 
+    uint256[] _attrs1,       // four times 
+    uint8[] _attrs2          // five times
```
Readability: "five times" you meant "base length x 5"? you have this check below, you can add the comments below. The comments here should be what the fields are for or meant, not the validation rules.

shaokun11 replied:
Thanks your good advice

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
+    for(uint8 i = 0; i < _isBitizenItems.length; i++) {
+      _tokenIds[i] = _mintToken(
+        _owner,
+        _addresses[i * 2 + 0],
```
no need to "+0"

shaokun11 replied:
Ok

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
+    emit CreateToken(_owner, _tokenId);
+  }
+  
+  function batchCreateToken(
```
Just curious: you can simply use this function for creating 1 token or N tokens. Why do you need to have both batch create and create?

shaokun11 replied:
Maybe I should remove it

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
+  }
+
+  function _convertToAttr1(uint8 _index, uint256[] _attrs1) private pure returns (uint256[4] _attr2) {
+    _attr2[0] = _attrs1[_index * 4 + 0];
```
Same as above, while I understand it looks formatted, but there's no need to do the "+ 0" here.

calvinmd commented on contracts/bitizens/AvatarItemOperator.sol:
```diff
@@ -1,48 +1,132 @@
+  }
+
+  function _convertToAttr2(uint8 _index, uint8[] _attrs2) private pure returns (uint8[5] _attr3) {
+    _attr3[0] = _attrs2[_index * 5 + 0]; 
```
Same as above

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+  struct Item {
+    address foundedBy;    // item founder
+    address createdBy;    // item creator
+    bool isBitizenItem;   // true for avatar false for other
```
very clear comments here, 👍 
if this is checking for Avatar, will it be clearer to just name it isAvatar?
isBitizenItem makes me think it should be true for an item, but false for avatar (the opposite)
Keeping as-is is fine with me, too.

shaokun11 replied:
My fault. It is real means  for bitizen,I'll fix it

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+  // tokenId => item
+  mapping(uint256 => Item) internal tokenInfo;
+  // all the burned token ids
+  uint256[] internal burnedTokenIds;
```
just curious, what is this for? why do you need to care for the burned tokens?

calvinmd replied:
Sure. If there's a use of it. Just beware of big arrays and the additional gas cost.

shaokun11 replied:
Now solidity save data just could use array or map. I think big array is also ok if we don't output all the data one time.

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+  // check token id => isBurned
+  mapping(uint256 => bool) internal isBurnedToken;
+  //hash(item) => tokenIds
+  mapping(bytes32 => uint256[]) internal sameItemTokenIds;
```
Would be great to have a brief description of what this mapping is for in the comment.
Why do you need to store an array of the same items?

shaokun11 replied:
For items deep have a summary some special item count

calvinmd replied:
If you only need a sum, would be sufficient to store a `mapping(byte32 => uint) sameTokenCount`?

shaokun11 replied:
I made an extension so that the same tokens could be found more easy.

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    _;
+  }
+
+  modifier existToken(uint256 _tokenId){
```
can we name it tokenExists()?

shaokun11 replied:
Yes,you are more professional on this field

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    avatarAccount = _account;
+  }
+
+  function compareItemSlots(uint256 _tokenId1, uint256 _tokenId2)
```
should it be named haveSameSockets()?

shaokun11 replied:
No,this name have a signature for avatar, we can't change it now

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    view 
+    existToken(_tokenId1)
+    existToken(_tokenId2)
+    returns (bool _res) {
```
If there is no meaning in the return variable name, you can remove the variable and just do return item1.sock == item2.socket below

shaokun11 replied:
Good way,I'll update it next time

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    return burnedTokenIds[_index];
+  }
+  
+  function getSameItemTokenIds(uint256 _tokenId) external view validToken(_tokenId) returns(uint256[]) {
```
Looks like a cool idea, would it be costly to maintain though? What if we have 100,000 users, and 10,000 of them have the same boot. How big will this array be?

shaokun11 replied:
Let me have a think of  this

shaokun11 replied:
I should use index to instead of directly return a array.

calvinmd replied:
👍

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
-  function getBurnTokenCount() external view returns (uint256) {
-    return burnTokenIds.length;
+  function isSameItem(uint256 _tokenId1, uint256 _tokenId2)
+  external
```
indentation. should have two leading spaces to be consistent

shaokun11 replied:
Yes, it should have two leading space

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    uint8[5] _attr2) 
+    external  
+    onlyOperator
+    returns(uint256 _tokenId) {
```
Lint: move { to the next line

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+    uint8 _energy,
+    uint8 _ext
+  ) 
+  external 
```
lint: add 2 spaces in front

calvinmd commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+  existToken(_tokenId) 
+  {
+    Item storage item = tokenInfo[_tokenId];
+    if (item.isBitizenItem != _isBitizenItem)
```
Is it necessary to check? Why not just assign the value in all cases?

shaokun11 replied:
Now I think it no necessary, haha

shaokun11 commented on contracts/bitizens/AvatarItemToken.sol:
```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+      item.probability,
+      _attr1,
+      _attr2
+      );
```
The same reason,solidity not allow a function show many variable,now I get information is the variable count is no more than 14 or 16

calvinmd replied:
Ah, got it. The stack too deep error. >_<

shaokun11 committed:
Simplify AvatarItemToken contract content and remove code that will not be used in the current phase.
Separate out AvatarItemPresale.sol from AvatarItemOperator,make the struct more clear

vernon99 reviewed (approved):
We basically have to rereview it all at that point. Lots of code that's hard to trace. Let's merge and do a last pass by the whole team if possible.

shaokun11 committed:
fine tuning some logic
Add car token

shaokun11 committed:
New version about avatar and item.
Avatar using the latest version ERC998
Add BitizenCar token
Corrected the names of some functions
